src/lib/shell/commands/Ota.cpp

```diff
@@ -0,0 +1,268 @@
+#if CHIP_DEVICE_LAYER_TARGET_NRFCONNECT
+#include "BDXDownloader_nrfconnect.h"
+#endif
+
+#include <controller-clusters/zap-generated/CHIPClientCallbacks.h>
+#include <controller-clusters/zap-generated/CHIPClusters.h>
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::Clusters::OtaSoftwareUpdateProvider::Commands;
+using namespace chip::bdx;
+using namespace chip::Callback;
+
+namespace chip {
+namespace Shell {
+namespace {
+
+constexpr EndpointId kProviderEndpointId                   = 0;
+constexpr const char kRequestorLocation[]                  = { 'U', 'S' };
+constexpr EmberAfOTADownloadProtocol kRequestorProtocols[] = { EMBER_ZCL_OTA_DOWNLOAD_PROTOCOL_BDX_SYNCHRONOUS };
+constexpr bool kRequestorCanConsent                        = false;
+
+BdxDownloader sBdxDownloader;
+Controller::DeviceControllerInteractionModelDelegate sIMDelegate;
+Transport::PeerAddress sProviderAddress;
+Shell::Engine sSubShell;
+
+CharSpan ExtractResourceName(CharSpan imageURI)
+{
+    // TODO: Implement a correct URI parser
+    constexpr char protocol[]    = "bdx://";
+    constexpr size_t protocolLen = sizeof(protocol) - 1;
+
+    if (imageURI.size() >= protocolLen && strncmp(protocol, imageURI.data(), imageURI.size()) == 0)
+    {
+        imageURI = imageURI.SubSpan(protocolLen, imageURI.size() - protocolLen);
+    }
+
+    for (size_t i = 0; i < imageURI.size(); ++i)
+    {
+        if (imageURI.data()[i] == '/')
+        {
+            return imageURI.SubSpan(i + 1, imageURI.size() - i - 1);
+        }
+    }
+
+    return imageURI;
+}
+
+void OnQueryImageResponse(void * context, const QueryImageResponse::DecodableType & response)
+{
+    ChipLogProgress(SoftwareUpdate, "Received QueryImage response: %" PRIu16, static_cast<uint16_t>(response.status));
+
+    CharSpan imageURI     = response.imageURI.ValueOr({});
+    CharSpan resourceName = ExtractResourceName(imageURI);
+    ChipLogProgress(SoftwareUpdate, "  Image URI: %.*s", static_cast<int>(imageURI.size()), imageURI.data());
+    ChipLogProgress(SoftwareUpdate, "  Resource: %.*s", static_cast<int>(resourceName.size()), resourceName.data());
+
+    OperationalDeviceProxy * deviceProxy = Server::GetInstance().GetOperationalDeviceProxy();
+    VerifyOrReturn(deviceProxy != nullptr);
+
+    Messaging::ExchangeManager * exchangeMgr = deviceProxy->GetExchangeManager();
+    Optional<SessionHandle> session          = deviceProxy->GetSecureSession();
+    Messaging::ExchangeContext * exchangeCtx = nullptr;
+
+    if (exchangeMgr != nullptr && session.HasValue())
+    {
+        exchangeCtx = exchangeMgr->NewContext(session.Value(), &sBdxDownloader);
+    }
+
+    if (exchangeCtx == nullptr)
+    {
+        ChipLogError(SoftwareUpdate, "Failed to allocate exchange");
+        return;
+    }
+
+    TransferSession::TransferInitData initOptions;
+    initOptions.TransferCtlFlags = TransferControlFlags::kReceiverDrive;
+    initOptions.MaxBlockSize     = 1024;
```

Comment:
consider making this some global constant so it can be used by `BDXDownloader` as well

Reply:
Done